Improve hit-rate on buildcaches

alalazo opened the pull request:
This PR tries to modify the optimization criteria to have more hits on buildcaches by default.

Modifications:
- [x] Assign compiler and target mismatches to the child node of the edge with a mismatch (previously was the parent node)
- [x] Move optimization of runtime nodes at very low priority, so that it doesn't interfere on whether specs are reused or not
- [x] Prevent specs from pre-`gcc-runtime` to be reused (they'll be mostly broken at runtime)
- [x] Bumped compiler versions in unit tests to avoid issues where a compiler doesn't support some newer targets (Apple M1)

alalazo committed:
Relax compiler and target mismatches

The mismatch occurs on an edge. Previously it was assigned
the parent priority, now it is assigned to the child
priority.

This should make reuse from buildcaches or store more likely,
since most mismatches will be counted with "reused" priority.

alalazo committed:
Optimize version badness for runtimes at very low priority

We don't want to e.g. switch other attributes because we
cannot reuse an old installed runtime.

haampie commented:
> to the child node

Smart

haampie commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3317,6 +3317,13 @@ def _is_reusable(spec: spack.spec.Spec, packages, local: bool) -> bool:
         return False
 
+    current_platform = spack.platforms.host()
+    if not spec.satisfies(f"platform={current_platform}"):
```
`satisfies` is a bit too much for this, plus it's a temporary Spec you're creating 1000s of times for large caches.

alalazo replied:
True, this needs to be optimized, and also the target part. I plan to do that and take care of the monkey patching issues in unit tests 🙂

alalazo replied:
I'll keep this for another PR.

alalazo committed:
Optimize runtime attributes at very low priority

This is such that the version of the runtime would
not influence whether we should reuse a spec.

Compiler mismatches are considered for runtimes,
to avoid situations where compiling foo%gcc@9
brings in gcc-runtime%gcc@13 if gcc@13 is among
the available compilers

alalazo commented:
Example of the current state (the occasional `icelake` nodes are the ones not in the buildcache):
![2024-03-20_22-59](https://github.com/spack/spack/assets/4199709/ce81f3cf-4842-4772-bc9b-aafd6188140e)

alalazo committed:
Add unit-tests to ensure better matching on reuse

alalazo commented:
The changes here give a much better hit rate on store / buildcaches. I tried to benchmark on a baseline _without_ any mirror and it't mostly the same. There's one consistent spike I need to investigate further:

[radiuss.pr.nomirror.csv](https://github.com/spack/spack/files/14704146/radiuss.pr.nomirror.csv)
[radiuss.develop.nomirror.csv](https://github.com/spack/spack/files/14704148/radiuss.develop.nomirror.csv)
[radiuss.txt](https://github.com/spack/spack/files/14704149/radiuss.txt)

![radiuss](https://github.com/spack/spack/assets/4199709/3566ac65-1118-48f4-ae9b-55d054f18cdd)

alalazo committed:
Exclude specs without runtimes from reuse

This should ensure that we do not reuse specs that
could be broken, as they expect the compiler to be
installed in a specific place.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3384,6 +3384,19 @@ def _is_reusable(spec: spack.spec.Spec, packages, local: bool) -> bool:
+    if not WITH_RUNTIME:
+        return True
+
+    if spec.satisfies("%gcc") and not spec.satisfies("^gcc-runtime"):
+        return False
+
+    if spec.satisfies("%oneapi") and not spec.satisfies("^intel-oneapi-runtime"):
+        return False
```
We might want to generalize this logic, and avoid hard-coding the current runtimes. This should be enough though to start testing the ASP logic, and see if it needs adjustments.

alalazo committed:
Bump clang version from 12 to 15 in unit-tests

This is to avoid differences on m1
with respect to x86_64 machines

alalazo committed:
Bump gcc version from 4.8.0 to 9.4.0 in unit tests

This is to avoid differences on m1
with respect to x86_64

alalazo commented:
I just tried to re-measure the benchmark above. There must be something in the `hiptt` spec that makes it very sensible to changes in the configuration[^1]. For what is worth, I still see a spike, but this time in the other direction. As before the difference always stems from the "solve" part.

* **Spack:** 0.22.0.dev0 (523c7042dcd48b599d1dffff89abb053024bd19f)
* **Python:** 3.11.5
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

[radiuss.pr.nomirror.csv](https://github.com/spack/spack/files/14838285/radiuss.pr.nomirror.csv)
[radiuss.develop.nomirror.csv](https://github.com/spack/spack/files/14838286/radiuss.develop.nomirror.csv)
[radiuss.txt](https://github.com/spack/spack/files/14838287/radiuss.txt)

![radiuss](https://github.com/spack/spack/assets/4199709/4fe63b93-88c3-4840-b3b2-ad0a0045b097)

[^1]: I still have to check what is the trigger. This time I have a few more compilers registered than in the test above.

haampie commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3394,6 +3394,19 @@ def _is_reusable(spec: spack.spec.Spec, packages, local: bool) -> bool:
+    if not WITH_RUNTIME:
+        return True
+
+    if spec.satisfies("%gcc") and not spec.satisfies("^gcc-runtime"):
```
```suggestion
    if spec.satisfies("%gcc") and not spec.dependencies("gcc-runtime"):
```

haampie replied:
don't traverse the tree while traversing the tree.

haampie replied:
(Actually it would be nice to have `has_dependency(...)` which doesn't allocate)

alalazo replied:
Done in https://github.com/spack/spack/pull/43272/commits/b5439b152c9faaa7c87a66c85e72be9de9b33cc8 I didn't do the non-allocating part, but fair to add in another PR.

haampie commented on lib/spack/spack/solver/asp.py:
```diff
@@ -3394,6 +3394,19 @@ def _is_reusable(spec: spack.spec.Spec, packages, local: bool) -> bool:
+    if spec.satisfies("%gcc") and not spec.satisfies("^gcc-runtime"):
+        return False
+
+    if spec.satisfies("%oneapi") and not spec.satisfies("^intel-oneapi-runtime"):
```
same

alalazo replied:
Done in https://github.com/spack/spack/pull/43272/commits/b5439b152c9faaa7c87a66c85e72be9de9b33cc8

haampie commented:
I think this is fine. As per discussion in dm, users can always `require: %gcc@13` to avoid incompatible mixing in case of fortran modules.

But we have to document it (in high impact changes for instance).

alalazo committed:
Improve check on dependencies

alalazo committed:
Improve check on compiler names
